### PR TITLE
H-2463, H-2499: Allow passing additional provenance data

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Makes it possible to provide custom provenance data for entities

## 🔍 What does this change?

- Populates the `provenance` field on entities, please see the changed files for more details,

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 1]
```diff
@@ -7,9 +7,167 @@ use bytes::BytesMut;
...
+    Flows,
+    Migration,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct OriginProvenance {
+    #[serde(rename = "type")]
+    pub ty: OriginType,
```
Comment (CiaranMn) [inline libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:136]:
```suggestion
    pub type: OriginType,
```

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 2]
```diff
@@ -7,9 +7,167 @@ use bytes::BytesMut;
...
+    /// The organization or company that produced the tool.
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub organization: Option<String>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct Artifact {
```
Comment (CiaranMn) [inline libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:41]: I would be tempted to leave this out until we know how exactly we will use it... we can discuss Monday

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs (hunk 1 above)
Comment (TimDiekmann) [reply libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:136]: `type` is not possible as it's a reserved keyword. It's (de-)serialized from/to `type`. This is common in Rust.

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 3]
```diff
@@ -7,9 +7,167 @@ use bytes::BytesMut;
...
+#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct Tool {
+    /// The name of the tool.
+    pub name: String,
+
+    /// The tool version, in whatever format the tool natively provides.
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub version: Option<String>,
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:25]: I would set/use the semantic version whenever it's possible/available, as they can be compared and compatibility can be checked. I don't think they need to match.

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs (hunk 2 above)
Comment (TimDiekmann) [reply libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs]: Removed in [`002ddaf` (#4306)](https://github.com/hashintel/hash/pull/4306/commits/002ddaf85c18c4f1284b156e56c487fd50cb1b61)

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 4]
```diff
@@ -7,9 +7,167 @@ use bytes::BytesMut;
...
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub organization: Option<String>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct Artifact {
+    /// The name of the tool.
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs]: Marking as resolved as it was removed in [`002ddaf` (#4306)](https://github.com/hashintel/hash/pull/4306/commits/002ddaf85c18c4f1284b156e56c487fd50cb1b61).

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs (hunk 1 above)
Comment (CiaranMn) [reply libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:104]: Ahh, sorry, didn't realise

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 5]
```diff
@@ -53,16 +179,29 @@ impl ToSql for EntityEditionProvenanceMetadata {
...
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub source: Vec<SourceProvenance>,
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub actor: Option<ActorType>,
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub tools: Vec<Tool>,
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub artifacts: Vec<Url>,
```
Comment (TimDiekmann) [inline libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:191]:
```suggestion
```

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 6]
```diff
@@ -7,9 +7,135 @@ use bytes::BytesMut;
...
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub enum OriginType {
+    WebApp,
+    MobileApp,
+    BrowserExtension,
+    Api,
+    Flows,
```
Comment (benwerner01) [inline libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:95]: Should this be singular?
```suggestion
    Flow,
```

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 7]
```diff
@@ -7,9 +7,135 @@ use bytes::BytesMut;
...
+    pub organization: Option<String>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub enum SourceType {
+    Webpage,
+    Document,
+}
```
Comment (TimDiekmann) [inline libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:44]: Discussed offline that we're keeping this as an enum and add fields as we need them. TBD: Add docs how to generate the OpenAPI specs for it.

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 8]
```diff
@@ -7,9 +7,135 @@ use bytes::BytesMut;
...
+#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct SourceProvenance {
+    #[serde(rename = "type")]
+    pub ty: SourceType,
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub authors: Vec<String>,
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub actor_type: Option<ActorType>,
```
Comment (TimDiekmann) [inline libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:56]:
```suggestion
```

File: libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs [hunk 9]
```diff
@@ -53,16 +176,27 @@ impl ToSql for EntityEditionProvenanceMetadata {
...
-    #[serde(default, rename = "__placeholder")]
-    __placeholder: (),
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub source: Vec<SourceProvenance>,
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub actor: Option<ActorType>,
+    #[cfg_attr(feature = "utoipa", schema(nullable = false))]
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub tools: Vec<Tool>,
```
Comment (TimDiekmann) [inline libs/@local/hash-graph-types/rust/src/knowledge/entity/provenance.rs:186]:
```suggestion
```

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**